pandas/core/internals/concat.py

```diff
@@ -219,6 +219,11 @@ def get_reindexed_values(self, empty_dtype, upcasted_na):
 
         else:
             for ax, indexer in self.indexers.items():
+                # GH-25014: get_upcasted_na returns iNaT, but
+                # DatetimeArray.take expects NaT.
+                # TODO: update get_empty_dtype_and_na to use EAs earlier?
+                if is_datetime64tz_dtype(values) and fill_value == tslibs.iNaT:
```

Comment:
isnull(fill_value)?

Reply:
actually this should be in take_nd I think (instead)

Reply:
I was (hopefully) able to fix it at the root, in `get_empty_dtype_and_na`.